### PR TITLE
Everywhere: Refactor version reporting so that the Kernel is the one source of truth

### LINKED ISSUE #15071: Rework (?) and unify versioning and its reporting in Serenity
Here's an overview what we currently do for version reporting:
- There are three pieces to the version: Major and minor (like semver), fixed to 1 and 0 respectively. The third part is the git commit hash (or part of it) which was the git HEAD during build.
- The image building process writes INI data of the following form into /res/version.ini:
```ini
[Version]
Major=1
Minor=0
Git=abcdef0
```
where the git hash is dynamically fetched from the environment by [build-root-filesystem.sh](https://github.com/SerenityOS/serenity/blob/27f5a18ce637598b83c5feec4e1392abe0100efc/Meta/build-root-filesystem.sh#L148).
- Core::Version reads this file to report a version of the form `Version {major}.{minor}.g{hash}`, which is used by AboutDialog and other parts of the GUI when asked about version information by the user.
- *However*, some parts of userland use the old hard-coded SERENITY_VERSION string from Core::Version, which even if only used as default arguments can slip into user-facing interfaces; as there is no discrepancy in practice the effects here are unknown.
- *However II*, `uname` (at least before #15068) reports hardcoded `1.0-dev` as POSIX's "revision" and `FIXME` as POSIX's "version". The latter is supposed to live below "revision", e.g. Linux reports the Kernel build timestamp, so the git commit hash makes a lot of sense there.
- The `uname` utility can't even output the version before #15068

As visible from this list, (1) there are significant inconsistencies between multiple version reporting paths and (2) as a first stepping stone #15068 brings Kernel version reporting through `uname` in line with Core::Version. As pointed out there, this leads to "weird" behaviour like the Kernel reading a file and parsing INI to retrieve version information. @linusg is not a fan and I am partially convinced.

I identify two paths forward, partially laid out by Linus and @nico:
1. Embed version information into the Kernel binary. Gets rid of strange kernel behaviour and establishes one clear path of version information retrieval: `uname(1)`. Is more complicated to implement (I for one have no idea how to incantate `objcopy` or Linker script for this).
2. Stick with the current behaviour but fully unify version reporting. Has the drawbacks of strange kernel behaviour.

And finally, there's the question of whether we want to use a build hash or a commit hash as the POSIX "version". The latter is much simpler. The former is more elegant but requires us to have reproducible builds.

Opinions welcome!

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
There's quite some stuff in this PR but it basically amounts to:
- Get rid of the three distinct version information sources there are right now (Core::Version::SERENITY_VERSION, /res/version.ini, uname syscall).
- Make the `uname` syscall the one source of truth for version information.
- Bake the actual version information into the Kernel; get rid of /res/version.ini.
- Make `uname(1)` (the utility) POSIX-compliant by reporting the git commit hash with `-v`.

image

Fixes #15071

### CODE REVIEW

**Review by linusg — COMMENTED** [review]
I personally hate the concept of a user-editable/deletable file somewhere on disk for version info (aka `/res/version.ini`), and using it in kernel land is even worse than in userland. Why can't we hardcode the version into the kernel at build time and expose it in `/proc` for userland?

**Comment by nico** [comment]
Embedding git revisions in binaries means that two identical source trees except for a comment change somewhere produce different binaries.

The cool way to handle this is to have deterministic builds and print the hash of the build outputs, not the inputs 😛

**Comment by linusg** [comment]
> two identical source trees except for a comment change somewhere produce different binaries.

In my mind that's an acceptable drawback / feature ;^), but I also understand the desire for reproducible builds. The problem of using the build output hash is that it's basically meaningless, compared to being able to trace a build back to a specific commit in the serenity repo.

**Comment by kleinesfilmroellchen** [comment]
> Embedding git revisions in binaries means that two identical source trees except for a comment change somewhere produce different binaries.

As long as you're not committing, the binaries will be the same. Either way, I don't know the best approach forward, but to me, it feels beside the point for this specific PR. We can have an additional discussion about how there's a hard-coded version string that some parts of userland use while Core::Version can already read /res/version just fine, but again, besides the point for this PR, :yaksteps: :^) What do you think, @linusg ?

**Comment by nico** [comment]
> it's basically meaningless, compared to being able to trace a build back to a specific commit in the serenity repo.

Au contraire! If your build is deterministic, the output hash clearly corresponds to one (or more! If there are multiple revs producing the same output) input rev.

**Comment by linusg** [comment]
Sure, but unless someone will be keeping a list of build hashes mapping to their corresponding commit hashes, I'm not sure how I would determine which "version" of serenity I'm running just from that?

File: Kernel/Syscalls/uname.cpp [hunk 1]
```diff
@@ -13,19 +16,67 @@ ErrorOr<FlatPtr> Process::sys$uname(Userspace<utsname*> user_buf)
...
+
+        AK::TypedTransfer<u8>::copy(reinterpret_cast<u8*>(buf.release), version_builder.to_byte_buffer().data(), min(version_builder.length(), UTSNAME_ENTRY_LEN - 1));
+
+        // Reuse the string builder for the version, i.e. the git commit hash.
+        version_builder.clear();
+
+        if (git_hash_location.has_value())
+            append_version_from_config(git_hash_location, 4);
+        else
+            version_builder.append("FIXME"sv);
```
Comment (linusg) [inline Kernel/Syscalls/uname.cpp:76]: Let's use "unknown" here

File: Kernel/Syscalls/uname.cpp [hunk 2]
```diff
@@ -13,19 +16,67 @@ ErrorOr<FlatPtr> Process::sys$uname(Userspace<utsname*> user_buf)
...
+    auto contents = StringView(raw_contents->bytes());
+
+    // FIXME: This INI parsing is super crude but we can't use Core::ConfigFile. As long as our version.ini doesn't do weird things, it should work.
+    // Find a way of reusing code across Kernel and Userland.
+    auto maybe_section_location = contents.find("[Version]\n"sv);
+    if (maybe_section_location.has_value()) {
+        auto section_location = maybe_section_location.value() + 10;
+        auto major_version_location = contents.find("Major="sv, section_location);
+        auto minor_version_location = contents.find("Minor="sv, section_location);
+        auto git_hash_location = contents.find("Git="sv, section_location);
```
Comment (linusg) [inline Kernel/Syscalls/uname.cpp:49]: I hadn't even seen this, now I dislike the idea of parsing version.ini in the kernel even more :sweat_smile:

**Review by linusg — COMMENTED** [review]
Big +1 for teaching uname about the hash as part of the version, but I'll need some other opinions about the current implementation (I'm biased, still).

File: Kernel/Syscalls/uname.cpp (hunk 1 above)
Comment (kleinesfilmroellchen) [reply Kernel/Syscalls/uname.cpp:76]: That would be a change in behaviour, I intentionally didn't want that if the git hash wasn't available, but sure.

File: Kernel/Syscalls/uname.cpp (hunk 2 above)
Comment (kleinesfilmroellchen) [reply Kernel/Syscalls/uname.cpp:49]: Let's see how the #15071 discussion goes.

File: Kernel/generate-version-file.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+#pragma once
+#include <AK/StringView.h>
+
+namespace Kernel {
+
+constexpr unsigned serenity_major_revision = 1;
+constexpr unsigned serenity_minor_revision = 0;
+constexpr StringView serenity_version = "${GIT_HASH}"sv;
```
Comment (linusg) [inline Kernel/generate-version-file.sh:20]: These should be ALL_CAPS like the old SERENITY_VERSION

**Review by linusg — COMMENTED** [review]
A file-based userspace API (i.e. /proc) would be nice, this breaks https://github.com/linusg/serenity_exporter/blob/26f0f049e5c25e34e6853549c55d5f010728f58b/src/collectors/os.js#L14-L16 unfortunately :upside_down_face:

Feel free to give it a try, otherwise I'll have to add it myself.

**Comment by kleinesfilmroellchen** [comment]
Ah, so that's why you need it. I just went with the minimal POSIX APIs we need to have.

File: Kernel/generate-version-file.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+
+set -e
+
+OUTPUT_FILE=$1
+GIT_HASH=$( (git log --pretty=format:'%h' -n 1 | cut -c1-7) || true )
```
Comment (GMTA) [inline Kernel/generate-version-file.sh:6]: Can we have a sensible default when `git` is not available, or the source code is not build with a `.git` directory present? (E.g. an export of the code)

Comment (GMTA) [reply Kernel/generate-version-file.sh:6]: Also: you might want to take into account whether we're building from unmodified source code, or a repository with outstanding changes (either in the working tree or index).

File: Kernel/Syscalls/uname.cpp [hunk 5]
```diff
@@ -13,19 +16,28 @@ ErrorOr<FlatPtr> Process::sys$uname(Userspace<utsname*> user_buf)
...
 #else
 #    error Unknown architecture
 #endif
+    };
+
+    StringBuilder version_builder;
+    version_builder.appendff("{}.{}-dev", serenity_major_revision, serenity_minor_revision);
+    AK::TypedTransfer<u8>::copy(reinterpret_cast<u8*>(buf.release), version_builder.to_byte_buffer().data(), min(version_builder.length(), UTSNAME_ENTRY_LEN - 1));
+
+    AK::TypedTransfer<u8>::copy(reinterpret_cast<u8*>(buf.version), serenity_version.bytes().data(), min(serenity_version.length(), UTSNAME_ENTRY_LEN - 1));
```
Comment (Hendiadyoin1) [inline Kernel/Syscalls/uname.cpp:40]: Lets do this properly and propagate errors, etc...
```suggestion
    KString version_string TRY(KString::formatted("{}.{}-dev", serenity_major_revision, serenity_minor_revision);
    AK::TypedTransfer<u8>::copy(buf.release, version_string.characters(), min(version_builder.length(), UTSNAME_ENTRY_LEN - 1));
    
    AK::TypedTransfer<u8>::copy(buf.version, serenity_version.characters_without_null_termination(), min(serenity_version.length(), UTSNAME_ENTRY_LEN - 1)); 
```

File: Userland/Libraries/LibCore/Version.cpp [hunk 6]
```diff
@@ -4,21 +4,23 @@
...
+
+    auto version = result.value().release;
+    auto git_hash = result.value().version;
 
     StringBuilder builder;
-    builder.appendff("Version {}.{}", major_version, minor_version);
-    if (auto git_version = version_config->read_entry("Version", "Git", ""); git_version != "")
-        builder.appendff(".g{}", git_version);
+    builder.appendff("Version {} revision {}", version, git_hash);
     return builder.to_string();
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibCore/Version.cpp:24]: This can be a String::formatted now

File: Kernel/Syscalls/uname.cpp [hunk 7]
```diff
@@ -13,23 +16,31 @@ ErrorOr<FlatPtr> Process::sys$uname(Userspace<utsname*> user_buf)
...
-    utsname buf {};
-    memcpy(buf.sysname, "SerenityOS", 11);
-    memcpy(buf.release, "1.0-dev", 8);
-    memcpy(buf.version, "FIXME", 6);
+    utsname buf
+    {
+        "SerenityOS",
+            "", // Hostname, filled in below.
+            "", // "Release" (1.0-dev), filled in below.
+            "", // "Revision" (git commit hash), filled in below.
```
Comment (ADKaster) [inline Kernel/Syscalls/uname.cpp:24]: constructing the buffer like this leaves the end of the buffers as uninitialized stack garbage and leaks information to userspace

Comment (kleinesfilmroellchen) [reply Kernel/Syscalls/uname.cpp:24]: Oh well, I would have guessed that some sensible defaults would zero a const char[] out, but C.

Comment (ADKaster) [reply Kernel/Syscalls/uname.cpp:24]: The old pattern does zero them out, by initializing with {} :)

File: Kernel/Syscalls/uname.cpp [hunk 8]
```diff
@@ -29,7 +29,7 @@ ErrorOr<FlatPtr> Process::sys$uname(Userspace<utsname*> user_buf)
 
     hostname().with_shared([&](auto const& name) {
         auto length = min(name->length(), UTSNAME_ENTRY_LEN - 1);
-        memcpy(buf.nodename, name->characters(), length);
+        AK::TypedTransfer<char>::copy(reinterpret_cast<char*>(buf.nodename), name->characters(), length);
```
Comment (Hendiadyoin1) [inline Kernel/Syscalls/uname.cpp:43]: shouldnt `nodename` decay to a `char*`. so that a cast is not needed, or do the templates mess with that

Comment (kleinesfilmroellchen) [reply Kernel/Syscalls/uname.cpp:43]: I tried, so no

**Comment by circl-lastname** [comment]
Is this still active?

**Comment by kleinesfilmroellchen** [comment]
I think it is totally mergeable, yes.

**Pull request merged**